zebra-consensus/src/block/subsidy/general.rs

```diff
@@ -25,38 +25,36 @@ pub fn halving_divisor(height: Height, network: Network) -> Option<u64> {
         .activation_height(network)
         .expect("blossom activation height should be available");
 
-    if height < SLOW_START_SHIFT {
+    let Some(height_after_slow_start_shift) = height - SLOW_START_SHIFT else {
         unreachable!(
             "unsupported block height: checkpoints should handle blocks below {:?}",
             SLOW_START_SHIFT
         )
-    } else if height < blossom_height {
-        let pre_blossom_height: u32 = (height - SLOW_START_SHIFT)
-            .try_into()
-            .expect("height is above slow start, and the difference fits in u32");
-        let halving_shift = pre_blossom_height / PRE_BLOSSOM_HALVING_INTERVAL.0;
-
-        let halving_div = 1u64
-            .checked_shl(halving_shift)
-            .expect("pre-blossom heights produce small shifts");
-
-        Some(halving_div)
-    } else {
-        let pre_blossom_height: u32 = (blossom_height - SLOW_START_SHIFT)
-            .try_into()
-            .expect("blossom height is above slow start, and the difference fits in u32");
-        let scaled_pre_blossom_height = pre_blossom_height * BLOSSOM_POW_TARGET_SPACING_RATIO;
-
-        let post_blossom_height: u32 = (height - blossom_height)
-            .try_into()
-            .expect("height is above blossom, and the difference fits in u32");
-
-        let halving_shift =
-            (scaled_pre_blossom_height + post_blossom_height) / POST_BLOSSOM_HALVING_INTERVAL.0;
+    };
 
-        // Some far-future shifts can be more than 63 bits
-        1u64.checked_shl(halving_shift)
-    }
+    let post_blossom_height_diff = (height - blossom_height)?;
+
+    post_blossom_height_diff
+        .is_negative()
+        .then(|| {
+            let pre_blossom_height = height_after_slow_start_shift.0;
+            let halving_shift = pre_blossom_height / (PRE_BLOSSOM_HALVING_INTERVAL as u32);
+
+            1u64.checked_shl(halving_shift)
+                .expect("pre-blossom heights produce small shifts")
+        })
+        .or_else(|| {
+            let pre_blossom_height = (blossom_height - SLOW_START_SHIFT)
+                .expect("Blossom height must be above SlowStartShift.")
+                .0;
+            let scaled_pre_blossom_height = pre_blossom_height * BLOSSOM_POW_TARGET_SPACING_RATIO;
+
+            let halving_shift = (scaled_pre_blossom_height + (post_blossom_height_diff as u32))
```

Comment:
This `as` is particularly risky because it operates on a number that's not a constant.

Reply:
155ce35d1d

Reply:
Turns out this is unnecessary with a765be43bf